src/core/boxed_packet.rs

```diff
@@ -0,0 +1,71 @@
+// Copyright 2022 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! Mixnet "boxed" packet types.
+
+use super::sphinx::{PeerId, PACKET_SIZE};
+
+#[derive(Clone)]
+/// Fixed-size "boxed" array, easily convertible to/from `Vec`.
+pub struct BoxedArray<T, const N: usize>(Vec<T>);
```

Comment:
Why not use `Box<[T]>`?  https://doc.rust-lang.org/std/vec/struct.Vec.html#method.into_boxed_slice

Reply:
Really just to allow easy conversion back to Vec; I couldn't see how to do this with Box.

Reply:
`.into()` works.

Reply:
Ah neat. Possibly this could even be `Box<[T; N]>`? I have just realised a different reason to prefer `Vec` though: it allows `truncate` to be implemented without a realloc and copy (`into_boxed_slice` will realloc and copy unless the `Vec` has no spare capacity). I'm currently using `truncate` to chop the `peel` output down to size before inserting it into the fragment assembler. There are other options here of course (eg just realloc and copy, or `peel` to a stack buffer first then move to a `BoxedArray` after). It's not clear which is the best really.

Reply:
I just made commands return their length and copied within the slice https://github.com/burdges/lake/blob/master/Xolotl/src/sphinx/layout.rs#L290 btw.

Reply:
I've gotten rid of `BoxedArray`, `BoxedPacket`, and `BoxedFragment`; seems possible and simpler to just use `Box<Packet>` and `Box<Fragment>`.